Sync/BatchingClient.swift

```
@@ -0,0 +1,227 @@
+        self.uploader = uploader
+        self.serializeRecord = serializeRecord
+
+        self.onCollectionUploaded = onCollectionUploaded
+    }
+
+    public func endBatch() -> Success {
+        guard !records.isEmpty else {
+            return succeed()
+        }
+
+        if let token = self.batchToken {
+            return commitBatch(token) >>> succeed
+        }
+
+        let lines = self.freezePost()
+        return self.uploader(lines: lines, ifUnmodifiedSince: self.ifUnmodifiedSince, queryParams: nil)
+            >>== effect(moveForward)
+            >>> succeed
+    }
+
+    public func addRecords(records: [Record<T>]) -> Success {
+        guard !records.isEmpty else {
+            return succeed()
+        }
+
+        do {
+            // Eagerly serializer the record prior to processing them so we can catch any issues
+            // with record sizes before we start uploading to the server.
+            let serialized = try records.map { try serialize($0) }
+            return addRecords(serialized.generate())
+        } catch let e {
+            return deferMaybe(e as! MaybeErrorType)
+        }
+    }
+
+    private func addRecords(generator: IndexingGenerator<[UploadRecord]>) -> Success {
+        var mutGenerator = generator
+        while let record = mutGenerator.next() {
+            return accumulateOrUpload(record) >>> { self.addRecords(mutGenerator) }
+        }
+        return succeed()
+    }
+
+    private func accumulateOrUpload(record: UploadRecord) -> Success {
+        do {
+            // Try to add the record to our buffer
+            try accumulateRecord(record)
+        } catch AccumulateRecordError.Full(let uploadOp) {
+            // When we're full, run the upload and try to add the record
+            // after uploading since we've made room for it.
+            return uploadOp >>> { self.accumulateOrUpload(record) }
+        } catch {
+            // Should never happen.
+            return succeed()
```

I'd return the error instead:

```
return deferMaybe(e)
```